# RQ-1: What source code properties characterize defective infrastructure as code scripts

## Mirantis

In [ ]:
import pandas as pd

mirantis = pd.read_csv("../data/IST_MIR.csv")

In [ ]:

X = mirantis[mirantis["defect_status"] == 0].drop("org", axis=1).drop("file_", axis=1)
Y = mirantis[mirantis["defect_status"] == 1].drop("org", axis=1).drop("file_", axis=1)
X


In [ ]:
import scipy
mann_pvalues = scipy.stats.mannwhitneyu(X, Y, method="exact", alternative="two-sided").pvalue

In [ ]:
from cliffs_delta import cliffs_delta # https://github.com/neilernst/cliffsDelta
results = {}
for index, column in enumerate(X):
    d, res = cliffs_delta(X[column], Y[column])
    results[column] = {
        "pvalue": mann_pvalues[index],
        "delta": d,
        "interpretation": res
    }
pd.DataFrame.from_dict(results)

### Mirantis results
|  | URL | File | Lines\_of\_code | Require | Ensure | Include | Attribute | Hard\_coded\_string | Comment | Command | File\_mode | SSH\_KEY | defect\_status |
| :--- | :--- | :--- | :--- | :--- | :--- | :--- | :--- | :--- | :--- | :--- | :--- | :--- | :--- |
| pvalue | 0.009196 | 0.000019 | 0.0 | 0.000032 | 0.000007 | 0.000145 | 0.0 | 0.0 | 0.000016 | 0.004841 | 0.000002 | 0.000004 | 0.0 |
| delta | -0.224702 | -0.364955 | -0.499752 | -0.355655 | -0.382316 | -0.325645 | -0.470238 | -0.551091 | -0.368552 | -0.242932 | -0.408606 | -0.392237 | -1.0 |
| interpretation | small | medium | large | medium | medium | small | medium | large | medium | small | medium | medium | large |


## Mozilla

In [107]:
import pandas as pd
import scipy
from cliffs_delta import cliffs_delta

mozilla = pd.read_csv("../data/IST_MOZ.csv")

X = mozilla[mozilla["defect_status"] == 0].drop("org", axis=1).drop("file_", axis=1)
Y = mozilla[mozilla["defect_status"] == 1].drop("org", axis=1).drop("file_", axis=1)


mann_pvalues = scipy.stats.mannwhitneyu(X, Y, method="exact", alternative="two-sided").pvalue

results = {}
for index, column in enumerate(X):
    d, res = cliffs_delta(X[column], Y[column])
    results[column] = {
        "pvalue": mann_pvalues[index],
        "delta": d,
        "interpretation": res
    }
pd.DataFrame.from_dict(results)

URL      File   Require      Ensure   Include  \
pvalue            0.080637  0.000135  0.000048    0.062343       0.0   
delta            -0.084329  -0.18368 -0.195612    -0.08997 -0.313331   
interpretation  negligible     small     small  negligible     small   

               Attribute Hard_coded_string   Command File_mode   SSH_KEY  \
pvalue               0.0               0.0   0.00016  0.000001  0.000001   
delta          -0.407558         -0.405333 -0.181672  -0.24097  -0.24097   
interpretation    medium            medium     small     small     small   

               Lines_of_code     Comment defect_status  
pvalue                   0.0    0.575592           0.0  
delta              -0.509725   -0.027063          -1.0  
interpretation         large  negligible         large

### Mozilla results
|  | URL | File | Require | Ensure | Include | Attribute | Hard\_coded\_string | Command | File\_mode | SSH\_KEY | Lines\_of\_code | Comment | defect\_status |
| :--- | :--- | :--- | :--- | :--- | :--- | :--- | :--- | :--- | :--- | :--- | :--- | :--- | :--- |
| pvalue | 0.080637 | 0.000135 | 0.000048 | 0.062343 | 0.0 | 0.0 | 0.0 | 0.00016 | 0.000001 | 0.000001 | 0.0 | 0.575592 | 0.0 |
| delta | -0.084329 | -0.18368 | -0.195612 | -0.08997 | -0.313331 | -0.407558 | -0.405333 | -0.181672 | -0.24097 | -0.24097 | -0.509725 | -0.027063 | -1.0 |
| interpretation | negligible | small | small | negligible | small | medium | medium | small | small | small | large | negligible | large |


## Openstack

In [ ]:
import pandas as pd
import scipy
from cliffs_delta import cliffs_delta

openstack = pd.read_csv("../data/IST_OST.csv")

X = openstack[openstack["defect_status"] == 0].drop("org", axis=1).drop("file_", axis=1)
Y = openstack[openstack["defect_status"] == 1].drop("org", axis=1).drop("file_", axis=1)


mann_pvalues = scipy.stats.mannwhitneyu(X, Y, method="exact", alternative="two-sided").pvalue

results = {}
for index, column in enumerate(X):
    d, res = cliffs_delta(X[column], Y[column])
    results[column] = {
        "pvalue": mann_pvalues[index],
        "delta": d,
        "interpretation": res
    }
pd.DataFrame.from_dict(results)


### Openstack results

## Wikimedia

In [1]:
import pandas as pd
import scipy
from cliffs_delta import cliffs_delta

wikimedia = pd.read_csv("../data/IST_WIK.csv")

X = wikimedia[wikimedia["defect_status"] == 0].drop("org", axis=1).drop("file_", axis=1)
Y = wikimedia[wikimedia["defect_status"] == 1].drop("org", axis=1).drop("file_", axis=1)


mann_pvalues = scipy.stats.mannwhitneyu(X, Y, method="exact", alternative="two-sided").pvalue

results = {}
for index, column in enumerate(X):
    d, res = cliffs_delta(X[column], Y[column])
    results[column] = {
        "pvalue": mann_pvalues[index],
        "delta": d,
        "interpretation": res
    }
pd.DataFrame.from_dict(results)

URL      File   Require    Ensure   Include Attribute  \
pvalue          0.010983  0.000003  0.000001  0.000018       0.0       0.0   
delta          -0.171429 -0.311387 -0.321417 -0.287509 -0.373545 -0.472648   
interpretation     small     small     small     small    medium    medium   

               Hard_coded_string   Command File_mode   SSH_KEY Lines_of_code  \
pvalue                       0.0  0.007938  0.000391  0.000277           0.0   
delta                  -0.546124 -0.178836 -0.238187 -0.244168     -0.513366   
interpretation             large     small     small     small         large   

                 Comment defect_status  
pvalue            0.0009           0.0  
delta          -0.223234          -1.0  
interpretation     small         large

### Wikimedia results
|  | URL | File | Require | Ensure | Include | Attribute | Hard\_coded\_string | Command | File\_mode | SSH\_KEY | Lines\_of\_code | Comment | defect\_status |
| :--- | :--- | :--- | :--- | :--- | :--- | :--- | :--- | :--- | :--- | :--- | :--- | :--- | :--- |
| pvalue | 0.010983 | 0.000003 | 0.000001 | 0.000018 | 0.0 | 0.0 | 0.0 | 0.007938 | 0.000391 | 0.000277 | 0.0 | 0.0009 | 0.0 |
| delta | -0.171429 | -0.311387 | -0.321417 | -0.287509 | -0.373545 | -0.472648 | -0.546124 | -0.178836 | -0.238187 | -0.244168 | -0.513366 | -0.223234 | -1.0 |
| interpretation | small | small | small | small | medium | medium | large | small | small | small | large | small | large |
